### PR TITLE
Ped animations interface

### LINKED ISSUE #187: 💡 [Animations]
**Is your feature request related to a problem? Please describe.**
👎

**Describe the solution you'd like**
   [Self]
         ↳ [Animations]

**Provide Reasoning**
They're always nice to have, for various reasons, and it would be safer and easier than loading an outdated SP ASI.

**Additional context**
_e.g. : Add a premade list, and/or "Play by name" + "Save & Load / Favorite" buttons._

### LINKED ISSUE #1624: [Request]: Add User-friendly Animations interface
### Problem

Currently animation plays only once within Debug-Animation tab. That would be great addition to make able to choise how to play animations: with checkbox to loop animation or to play it once if it unchecked. 

### Solution

This request is for small but quite useful feature to add a checkbox that a loop animation selected and Fetched from Debug.

### Reason

Some animations are made to be looped, like dances or idles and playing them once is look so crumplet.

### Additional context

[EDIT](https://github.com/YimMenu/YimMenu/issues/1624#issuecomment-1623872015): request for user-friendly interface for animations but don't remove debug feature yet.
There is an animations LUA (Author: "DEEZY", he rewrite this from FiveM) that uses most useful animations and part of them can play them with props attached to player model bones [Animations2.zip](https://github.com/YimMenu/YimMenu/files/11970128/Animations2.zip). That file may can help to make new feature as code reference in future release of user-friendly interface. 

We must have user-friendly animations to able to: 
Use animations without locking player movement (urgent in certain animations) to animate (upper) body part with able to move freely while animation is playin;
Would be great to use animations with props attached and safely remove them after stoping playin or with special "Deattach" button (make it to choose maybe);
Able to loop animations, stop them or play them once;
Interface of user-friendly may be separated to different categories of animations (maybe droplist or menu select) and they can be like: Idle, Sitting, Romance, Emotes (Paired, Actions, Taunts, Sport, Car, etc), With Props (guitars, guns, everythin else), Animals animations when transformed;
Prop spawn feature above can be a part of code for Prop-spawn-user-friendly-interface that we also in need.

### PR DESCRIPTION
**Pull request by Tayipc38**
Created a new service to save and make new animation entries. For now I've integrated Debug Animations in to have at least one source to fetch from. This does mean users would have to download [DurtyFree animDictsCompact.json](https://github.com/DurtyFree/gta-v-data-dumps/blob/master/animDictsCompact.json) for themselves. Otherwise a data service addition is in order to fetch animations.

Needs translation for Tab label.

Closes #187
Closes #1624 
Does not have the attached props part.

image

Added ambient animations that can be played depending on saved coordinates and rotation data. Player will be prompted when close:
image
image

The `DANCE` operation in the context menu that is reserved for non-player peds, has been altered to play the animation that is currently deployed in the animations UI under Self. The operation is renamed as `ANIMATION`. In case an invalid dict is detected, the operation will default to the dance. Ambient animations will also work on peds, that way a "scene" of some sort can be created.

### CODE REVIEW

**Comment by alyxme** [comment]
> * [binary.zip](https://nightly.link/YimMenu/YimMenu/actions/artifacts/795008183.zip)

Clicking on Debug-Animation tab cause crash without any info about that at cout.log

**Comment by Tayipc38** [comment]
> > * [binary.zip](https://nightly.link/YimMenu/YimMenu/actions/artifacts/795008183.zip)
> 
> Clicking on Debug-Animation tab cause crash without any info about that at cout.log

Correct. Ill fix it in a second. I was derefencing null pointers 🤦.

**Comment by Yimura** [comment]
This PR should remove the need for #1667

**Comment by alyxme** [comment]
> This PR should remove the need for #1667

It is ok! I've spent several hours to do that. Now I understand more. Just remove that cos @DayibBaba did a great new feature we all need!

**Comment by Tayipc38** [comment]
> > This PR should remove the need for #1667
> 
> It is ok! I've spent several hours to do that. Now I understand more. Just remove that cos @DayibBaba did a great new feature we all need! 

Ill use Ai to convert that animations lua you had posted in an issue.

**Comment by muratcansimsek** [comment]
Can I contact you privately? @DayibBaba 
dc: muratcansimsek

File: src/util/animations.hpp [hunk 1]
```diff
@@ -59,7 +93,7 @@ namespace big::animations
 		{
 			LOG(INFO) << "Animations file is not in directory. https://raw.githubusercontent.com/DurtyFree/gta-v-data-dumps/master/animDictsCompact.json";
```
Comment (Yimura) [inline src/util/animations.hpp]: Make the log WARNING/ERROR to make it more obvious to users, they may not notice if the log level is just info.

File: src/views/debug/view_debug.hpp [hunk 2]
```diff
@@ -9,7 +9,7 @@ namespace big::debug
 	extern void script_events();
 	extern void scripts();
 	extern void threads();
-	extern void animations();
+	extern void animations(std::string* dict = nullptr, std::string* anim = nullptr); // Can be used to retrieve animations
```
Comment (Yimura) [inline src/views/debug/view_debug.hpp:12]: Function params can be replaced by `std::string& dict, std::string& anim`.

An empty string can then be used to indicate the default state.
```cpp
if (!empty(dict) && !empty(anim))
{
    // code
}
```

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/views/debug/view_debug.hpp (hunk 2 above)
Comment (Tayipc38) [reply src/views/debug/view_debug.hpp:12]: How about calling `animations` without any strings? For example the original call to render it in the debug window. I originally had them as references but decided to make them pointers for this reason.

**Comment by lonelybud** [comment]
image
image
image

using this conifg to achieve final state

**Comment by lonelybud** [comment]
great UI. this is what i was a ble to create  `ped_animations.json` - 
[ped_animations.zip](https://github.com/YimMenu/YimMenu/files/12041764/ped_animations.zip)

**Comment by lonelybud** [comment]
when can we expect this to merge @DayibBaba

**Comment by Yimura** [comment]
> when can we expect this to merge

There will be no ETA given. Please be patient.

**Comment by lonelybud** [comment]
https://github.com/Svxy/TnyavnTos-GTA5-Mod-Menu

support for props during animation is present in this menu.

**Comment by SkiddyToast** [comment]
> https://github.com/Svxy/TnyavnTos-GTA5-Mod-Menu
> 
> support for props during animation is present in this menu.

Those are called scenarios.

**Comment by lonelybud** [comment]
yes. only few seems to work properly in online. Also it has few animations as well (11 total)
image

**Comment by Tayipc38** [comment]
Props for animations apart from scenarios might be something ill add later on, for now I aim to finish this PR by the end of next week with one more addition to it.

**Comment by dynamoNg** [comment]
here i did something special. I shared the variable used by animation service with the `context menu > dance (ped)`. This will allow any animation to be played on the ped.

Uploading Grand Theft Auto V 2023-07-31 02-05-39.mp4…

**Comment by lonelybud** [comment]
here i did something special. I shared the variable used by animation service with the `context menu > dance (ped)`. This will allow any animation to be played on the ped.

https://github.com/YimMenu/YimMenu/assets/139460769/f75fb64e-cc54-4a3d-a5f9-3deb3d9281b7

**Comment by lonelybud** [comment]
> here i did something special. I shared the variable used by animation service with the `context menu > dance (ped)`. This will allow any animation to be played on the ped.
> 
>  Grand.Theft.Auto.V.2023-07-31.02-05-39.mp4

https://github.com/dynamoNg/YimMenu/pull/5/commits/dca20376c867e8fbd81a8689e737e69288049cd7
https://github.com/dynamoNg/YimMenu/actions/runs/5708308246/job/15465852573

it will be so nice if something like this can be added

**Comment by lonelybud** [comment]
i tried the scenarios (https://github.com/DurtyFree/gta-v-data-dumps/blob/master/scenariosCompact.json)

Added scenario in animation menu at bottom just for fun, also shared state with context menu. 
https://github.com/dynamoNg/YimMenu/actions/runs/5730675112

it worked as follows - 
- for male player / npc expected animations like playing drum, drinking coffee worked well.
- while for female players / npc some animations like mentioned above did not worked.

**Comment by lonelybud** [comment]
only some scenarios work on female character as illustrated in this video. So overall scenarios are working fine as expected.
https://www.youtube.com/watch?v=k14GP5QHS1E

**Comment by Tayipc38** [comment]
> https://github.com/dynamoNg/YimMenu/commit/dca20376c867e8fbd81a8689e737e69288049cd7
https://github.com/dynamoNg/YimMenu/actions/runs/5708308246/job/15465852573

>it will be so nice if something like this can be added

I've added this as well, but no need for an inline variable as the class instance is a global itself. With this added, I think the initial feature is ready. Some additions and improvements can be added down the line (such as scenarios).

File: src/main.cpp [hunk 3]
```diff
@@ -127,6 +127,7 @@ BOOL APIENTRY DllMain(HMODULE hmod, DWORD reason, PVOID)
 			    g_script_mgr.add_script(std::make_unique<script>(&context_menu_service::context_menu, "Context Menu"));
 			    g_script_mgr.add_script(std::make_unique<script>(&backend::tunables_script, "Tunables"));
 			    g_script_mgr.add_script(std::make_unique<script>(&backend::squad_spawner, "Squad Spawner"));
+				g_script_mgr.add_script(std::make_unique<script>(&backend::ambient_animations_loop, "Ambient Animations"));
```
Comment (Yimura) [inline src/main.cpp]: Different indentation from the other lines.

File: src/views/debug/view_debug_animations.cpp [hunk 4]
```diff
@@ -5,72 +5,72 @@
...
-				selected_dict_anim_list.clear();
-				auto range = animations::all_anims.equal_range(current_dict);
-				for (auto it = range.first; it != range.second; ++it)
-				{
-					selected_dict_anim_list.push_back(it->second);
-				}
-			};
+		if(dict && anim)
+			*dict = current_dict,
+			*anim = current_anim;
```
Comment (Yimura) [inline src/views/debug/view_debug_animations.cpp]: Just use normal curly brackets
```cpp
if ()
{

}
```

**Review by Yimura — CHANGES_REQUESTED** [review]

**Comment by lonelybud** [comment]
> I've added this as well, but no need for an inline variable as the class instance is a global itself. With this added, I think the initial feature is ready. Some additions and improvements can be added down the line (such as scenarios).

thanks i will check. I have checked [scenarios](https://github.com/dynamoNg/YimMenu/actions/runs/5739857052) as well they are working fine. It will added later on.

https://github.com/YimMenu/YimMenu/assets/139460769/c40c23bc-be7c-4e36-90fd-3f952bce4330

**Review by Yimura — APPROVED** [review]

**Pull request merged**